src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.core;
+
+import io.strimzi.kafka.bridge.mqtt.utils.MappingRule;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+/**
+ * Responsible for handling all the topic mapping.
+ *
+ * @see MappingRule
+ */
+public class MqttKafkaMapper {
+
+    private static String DEFAULT_KAFKA_TOPIC = "messages_default";
+    private ArrayList<MappingRule> rules;
+    // matches any character except line terminators. Used to replace the placeholders with {something} in the mqtt pattern.
+    private static String PLACEHOLDER_REGEX = "(.*)";
+    // matches any character except a forward slash (/). Used to replace the + in the mqtt pattern.
+    private static String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+
+    /**
+     * Constructor
+     * <p>
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(ArrayList<MappingRule> rules) {
+        this.rules = rules;
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            //convert the mqtt pattern to a valid regex expression.
+            // the mqtt pattern can contain placeholders like {something}, + and #.
+            // if the mqtt topic contains a +, we replace it with @singleLevelWildcardRegex
+            // if the mqtt topic contains a #, we replace it with @multiLevelWildcardRegex
+            // if the mqtt topic contains a placeholder (pattern \{\w+\}), we replace it with @placeholderRegex
+            String regex = rule.getMqttTopicPattern().replaceAll("\\{\\w+\\}", PLACEHOLDER_REGEX)
+                    .replace("#", MULTIPLE_LEVEL_WILDCARD_REGEX).replace("+", SINGLE_LEVEL_WILDCARD_REGEX);
```

Comment:
Would it make sense to translate the MappingRules to regex outside of the method so we only have to do it once instead of every time we have to map a mqttTopic name? Maybe we could do it as part of the MappingRules class?

Reply:
Totally agree with Kyle. On class creation we should "map" the rules from JSON to our internal list of regex.
It would be better to use the `Pattern` class and the `Pattern.compile` static method. 
You could "compile" each regex and having the corresponding list of `Pattern` instances.
This way the matching operation would be even faster.

So you could have a `List<Pattern> patterns` field and filling it going through the rules, compiling each regex.

```java
for (MappingRule rule : this.rules) {
    String regex = rule.getMqttTopicPattern().replaceAll("\\{\\w+\\}", PLACEHOLDER_REGEX)
                      .replace("#", MULTIPLE_LEVEL_WILDCARD_REGEX).replace("+", SINGLE_LEVEL_WILDCARD_REGEX);
    patterns.add(Pattern.compile(regex));
}
```

Reply:
Let's see how it is :-)

Reply:
@ppatierno you can take a look now.